blob/gcsblob/gcsblob.go

```diff
@@ -70,48 +54,85 @@ import (
 const defaultPageSize = 1000
 
 func init() {
-	blob.Register("gs", openURL)
+	blob.DefaultURLMux().RegisterBucket(Scheme, new(lazyCredsOpener))
 }
 
-func openURL(ctx context.Context, u *url.URL) (driver.Bucket, error) {
-	q := u.Query()
-	opts := &Options{}
-
-	if accessID := q["access_id"]; len(accessID) > 0 {
-		opts.GoogleAccessID = accessID[0]
-	}
+// lazyCredsOpener obtains Application Default Credentials on the first call
+// to OpenBucketURL.
+type lazyCredsOpener struct {
+	init   sync.Once
+	opener *URLOpener
+	err    error
+}
 
-	if keyPath := q["private_key_path"]; len(keyPath) > 0 {
-		pk, err := ioutil.ReadFile(keyPath[0])
+func (o *lazyCredsOpener) OpenBucketURL(ctx context.Context, u *url.URL) (*blob.Bucket, error) {
+	o.init.Do(func() {
+		creds, err := gcp.DefaultCredentials(ctx)
 		if err != nil {
-			return nil, err
+			o.err = err
+			return
 		}
-		opts.PrivateKey = pk
-	}
-
-	var creds *google.Credentials
-	if credPath := q["cred_path"]; len(credPath) == 0 {
-		var err error
-		creds, err = gcp.DefaultCredentials(ctx)
+		client, err := gcp.NewHTTPClient(gcp.DefaultTransport(), creds.TokenSource)
 		if err != nil {
-			return nil, err
+			o.err = err
+			return
 		}
-	} else {
-		jsonCreds, err := ioutil.ReadFile(credPath[0])
-		if err != nil {
-			return nil, err
+		o.opener = &URLOpener{Client: client}
+	})
+	if o.err != nil {
+		return nil, fmt.Errorf("open GCS bucket %q: %v", u, o.err)
+	}
+	return o.opener.OpenBucketURL(ctx, u)
+}
+
+// Scheme is the URL scheme gcsblob registers its URLOpener under on
+// blob.DefaultMux.
+const Scheme = "gs"
+
+// URLOpener opens GCS URLs like "gs://mybucket".
+//
+// This opener supports the following query parameters:
+//
+//     access_id: sets Options.GoogleAccessID
+//     private_key_path: path to read for Options.PrivateKey
+type URLOpener struct {
+	// Client must be set to a non-nil HTTP client authenticated with
+	// Cloud Storage scope or equivalent.
+	Client *gcp.HTTPClient
+
+	// Options specifies the default options to pass to OpenBucket.
+	Options Options
+}
+
+// OpenBucketURL opens the GCS bucket with the same name of the URL's host.
+func (o *URLOpener) OpenBucketURL(ctx context.Context, u *url.URL) (*blob.Bucket, error) {
+	var err error
+	o, err = o.forParams(ctx, u.Query())
```

Comment:
It took me a while to parse what's going on here (it's kind of weird to overwrite `o`). Could it just be done inline, and only copying/update `Options` instead of a new `URLOpener`?